Add the Jobs SDK

siri-varma opened the pull request:
# PR Overview: Jobs SDK Integration

This PR introduces the Jobs SDK with three main APIs: **Schedule**, **Get**, and **Delete Job**. The implementation is divided into the following categories:

- **Core Logic**
- **Models**
- **Tests (Unit and Integration)**

---

## Core Logic

The core logic is implemented in two key classes, as detailed below:

| **Class**                 | **Description**                                                                                       | **Details**                                                                                                         |
|---------------------------|-------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------|
| **CronExpressionBuilder**  | Builds Cron expressions compliant with Dapr sidecar expectations.                                     | - Supports valid Cron syntax (`*`, `/`, and other ranges). <br> - References: [Jobs API](https://docs.dapr.io/reference/api/jobs_api/#schedule) and .NET SDK. |
| **DaprJobsClient**         | Manages the ScheduleJob, Get, and DeleteJob APIs. Handles validations and async calls.               | - Schedules, retrieves, and deletes jobs. <br> - Validates inputs and performs async operations for job management.  |

---

## Models

The rest of the classes in this SDK are models that are constructed using the **builder pattern**, ensuring clean and maintainable object creation.

---

## Tests

The testing section includes:

- **Unit Tests**: Focus on validating the functionality of individual components.
- **Integration Tests**: Ensure the system's interaction with external dependencies (e.g., the Dapr sidecar) works as expected.

---

## Issue Reference

We strive to ensure that all PRs are opened based on a relevant issue, where the problem or feature has been discussed prior to implementation.

This PR addresses and closes the following issue:  
**#1101**

---

## Checklist

Please make sure you've completed the relevant tasks for this PR:

- [X] Code compiles correctly
- [X] Created/updated tests
- [X] Extended the documentation

---

This structure organizes the content clearly and makes it easy to digest, with section headers, tables for the core logic, and a clean checklist at the end.

salaboy commented on sdk-jobs/pom.xml:
```diff
@@ -0,0 +1,154 @@
+<project
```
@siri-varma do we really need a new maven module and dependency? Workflows was separated because the SDK is doing quite a lot, the Conversation API is only consuming a sidecar API and this one too.. i am not entirely sure that we should create whole new modules for just adding APIs. We don't have a state store module or a pubsub one.

siri-varma replied:
My understanding was that we are going more modular. That is what I have seen with the .NET SDK . If this is not the case that is fine as well.

On a personal note, I also prefer modular packages because they offer the flexibility to include only what you need.

Let me know if this needs further discussion or if we want to stick with singular module

WhitWaldo replied:
It's up to each SDK what they want to do. 

In the JS SDK, they're all one package, but separate actual clients (as tree shaking during a standard build will remove anything not used).

In the .NET SDK, each building block is getting a separate package for ease of maintenance, to take an opportunity to remove and refactor obsolete code, and facilitate a better development experience.

Neither of those necessarily apply to the Java SDK as that's up to these maintainers.

siri-varma replied:
Yep. This was a learning for me.

salaboy commented on sdk-jobs/src/main/java/io/dapr/jobs/client/DayOfWeek.java:
```diff
@@ -0,0 +1,39 @@
+ * System.out.println(day.getRank()); // Outputs: 1
+ * </pre>
+ */
+public enum DayOfWeek implements OrdinalEnum {
```
@siri-varma maybe I am lacking context here.. but shouldn't we rely on a more standard Java way of doing this? If we define our own DayOfWeek Enum, then people will have to map to whatever they are using for their calendars.

siri-varma replied:
https://docs.dapr.io/reference/api/jobs_api/#schedule

Month, DayOfWeek are to be used only for building CronExpression. And the syntax for CronExpression is different from the regular Java date time constructs. For eg: Month in Java has JANUARY while Cron has Jan. 

I think the confusion is coming from how we named the class. I can rename them to be more specific like CronDayOfWeek, CronMonthOfYear.

salaboy commented on sdk-jobs/src/main/java/io/dapr/jobs/client/MonthOfYear.java:
```diff
@@ -0,0 +1,38 @@
+ * </pre>
+ */
+
+public enum MonthOfYear implements OrdinalEnum {
```
@siri-varma same here.. I would rather use standard dates to define the months.

siri-varma replied:
Same as here https://github.com/dapr/java-sdk/pull/1255#discussion_r1995847248

salaboy commented on sdk-jobs/src/main/java/io/dapr/jobs/client/OrdinalEnum.java:
```diff
@@ -0,0 +1,16 @@
+ * This interface is intended to be implemented by enums that need to provide
+ * a numerical representation for their values, such as days of the week or months of the year.
+ */
+public interface OrdinalEnum {
```
More like this, using  `DateTimeFormatter` and `LocalDate`

```
public class DayOfWeekConverter {

    public static String getDayOfWeekShort(String dateInput) {
        DayOfWeek dayOfWeek = getDayOfWeek(dateInput);
        return dayOfWeek.getDisplayName(TextStyle.SHORT, Locale.US);
    }

    private static DayOfWeek getDayOfWeek(String dateInput) {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/M/yyyy");
        LocalDate localDate = LocalDate.parse(dateInput, formatter);
        return localDate.getDayOfWeek();
    }

    public static int getDayOfWeekOrdinal(String dateInput) {
        DayOfWeek dayOfWeek = getDayOfWeek(dateInput);
        return dayOfWeek.getValue();
    }
}
```

siri-varma replied:
Same as here https://github.com/dapr/java-sdk/pull/1255#discussion_r1995847248

salaboy commented on sdk-jobs/src/main/java/io/dapr/jobs/client/CronPeriod.java:
```diff
@@ -0,0 +1,44 @@
+ * System.out.println(period); // Outputs: MINUTES
+ * </pre>
+ */
+public enum CronPeriod {
```
@siri-varma similar here.. in Spring Framework they already have this.. but can we not use plain java constructs here?

siri-varma replied:
Couple of questions

* Is there a possibility that users would not want to use spring ?
* Also, spring's cron supports lot of other things than what dapr supports (https://docs.dapr.io/reference/api/jobs_api/#schedule).

hadeed85 replied:
I don't mean to interject, but some of us are waiting on the Jobs API and would not like to use Spring.
Case in point: https://github.com/dapr/java-sdk/issues/976

salaboy replied:
I wasn't suggesting to depend on spring but also not to create internal representations that then we will need to map for spring users. Also there are pure Java solutions too

salaboy commented on testcontainers-dapr/src/main/java/io/dapr/testcontainers/DaprSchedulerContainer.java:
```diff
@@ -0,0 +1,82 @@
+/**
+ * Test container for Dapr scheduler service.
+ */
+public class DaprSchedulerContainer extends GenericContainer<DaprSchedulerContainer> {
```
@siri-varma nice!!!

salaboy reviewed (commented):
@siri-varma can you take a look at my comments? @artur-ciocanu I am curious about your feedback here too.

artur-ciocanu commented:
@siri-varma thanks for all your work! I have reviewed the PR and I wanted to share a few general observations/recommendations:
- As far as I saw different maintainers decided to approach modularity differently. In Dotnet SDK we have separate DLLs for different functionality, in Python everything is in a single client, but the APIs are suffixed with `alpha1`. In Java SDK we have a special interface called `DaprPreviewClient` that should host all the alpha APIs. I would recommend having a single client and keep everything in `sdk` module and use `DaprPreviewClient` for alpha APIs like Jobs API and Conversation API
- Similar to above, let's push everything to `sdk` module, so we avoid creating additional Maven modules
- Since Jobs API is still alpha, I think it is too early to commit to creating our own CRON expression builder. As far as I can tell a lot of code in this PR is dedicated to CRON expressions. For now I think it is enough to just use CRON expressions as string value. The only thing that we should make sure is that the CRON expression is valid, although I think if it is not, the Dapr sidecar will complain.
- In the Jobs API that have been added there is a lot of repetition regarding validating model fields like something is not null etc, I would suggest extracting all those validations to separate methods like:
  - `validateScheduleJobRequest(...)` - to validate `ScheduleJobRequest`
  - `validateGetJobRequest(...)` - to validate `GetJobRequest`
  - `validateDeleteJobRequest(...)` - to validate `DeleteJobRequest`
This way we will have clean separation between data validation and GRPC method invocation.
- You might want to look into other APIs to make sure that you use the right interceptors to ensure tracing data is properly propagated. Although we could address this later, as well.

CC: @salaboy @cicoyle

artur-ciocanu reviewed (commented):
@siri-varma please check my comment and let me know your thoughts.

siri-varma commented:
Had a discussion with @artur-ciocanu regarding the comments and this is what we were thinking 

> You might want to look into other APIs to make sure that you use the right interceptors to ensure tracing data is properly propagated. Although we could address this later, as well.
> The only thing that we should make sure is that the CRON expression is valid

Will address the above in a different PR

Will move CronExpressionBuilder to a different PR for two reasons
* Because of the amount of logic present in the builder
* We also want to discuss how the SDK should support such a builder while considering both plain Java and Spring implementations.

cc: @salaboy / @cicoyle

siri-varma commented:
@artur-ciocanu Addressed all the comments

siri-varma commented:
> @siri-varma thanks for all your work! I have reviewed the PR and I wanted to share a few general observations/recommendations:
> 
> * As far as I saw different maintainers decided to approach modularity differently. In Dotnet SDK we have separate DLLs for different functionality, in Python everything is in a single client, but the APIs are suffixed with `alpha1`. In Java SDK we have a special interface called `DaprPreviewClient` that should host all the alpha APIs. I would recommend having a single client and keep everything in `sdk` module and use `DaprPreviewClient` for alpha APIs like Jobs API and Conversation API
> * Similar to above, let's push everything to `sdk` module, so we avoid creating additional Maven modules
> * Since Jobs API is still alpha, I think it is too early to commit to creating our own CRON expression builder. As far as I can tell a lot of code in this PR is dedicated to CRON expressions. For now I think it is enough to just use CRON expressions as string value. The only thing that we should make sure is that the CRON expression is valid, although I think if it is not, the Dapr sidecar will complain.
> * In the Jobs API that have been added there is a lot of repetition regarding validating model fields like something is not null etc, I would suggest extracting all those validations to separate methods like:
>   
>   * `validateScheduleJobRequest(...)` - to validate `ScheduleJobRequest`
>   * `validateGetJobRequest(...)` - to validate `GetJobRequest`
>   * `validateDeleteJobRequest(...)` - to validate `DeleteJobRequest`
>     This way we will have clean separation between data validation and GRPC method invocation.
> * You might want to look into other APIs to make sure that you use the right interceptors to ensure tracing data is properly propagated. Although we could address this later, as well.
> 
> CC: @salaboy @cicoyle

Addressed the comments

siri-varma commented:
/assign

artur-ciocanu commented on sdk-tests/src/test/java/io/dapr/it/testcontainers/DaprJobsIT.java:
```diff
@@ -0,0 +1,132 @@
+  private static final Random RANDOM = new Random();
+  private static final int PORT = RANDOM.nextInt(1000) + 8000;
+
+  @Container
```
Here we should probably use the DaprConstants IMAGE_TAG

siri-varma replied:
Changed to use image tag

artur-ciocanu commented on sdk-tests/src/test/java/io/dapr/it/testcontainers/DaprJobsIT.java:
```diff
@@ -0,0 +1,132 @@
+    daprPreviewClient.scheduleJob(new ScheduleJobRequest("Job", currentTime)).block();
+
+    GetJobResponse getJobResponse =
+        daprPreviewClient.getJob(new GetJobRequest("Job")).block();
```
A minor nitpick, for readability and since we are in Unit Test land I would recommend using static imports so we could use `asertEquals`, etc directly.

The same applies to other Assertions.

siri-varma replied:
Changed to assertEquals with static imports

artur-ciocanu commented on sdk-tests/src/test/java/io/dapr/it/testcontainers/TestDaprJobsConfiguration.java:
```diff
@@ -0,0 +1,40 @@
+  public DaprPreviewClient daprPreviewClient(
+      @Value("${dapr.http.endpoint}") String daprHttpEndpoint,
+      @Value("${dapr.grpc.endpoint}") String daprGrpcEndpoint
+  ){
```
It is a little bit unclear why `overrides` variable is not used.

siri-varma replied:
Using overrides

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprClientHttpTest.java:
```diff
@@ -13,8 +13,7 @@
 package io.dapr.client;
 
 import com.fasterxml.jackson.core.JsonParseException;
```
Please configure your IDE to always use full imports.

siri-varma replied:
Used full imports

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprClientHttpTest.java:
```diff
@@ -45,14 +44,9 @@
-import static org.junit.jupiter.api.Assertions.assertNull;
-import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.*;
 import static org.mockito.ArgumentMatchers.any;
```
Same here, let's use full imports

siri-varma replied:
used full imports

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -16,34 +16,33 @@
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.protobuf.Any;
 import com.google.protobuf.ByteString;
```
Please use full imports

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -57,16 +56,10 @@
 import java.util.concurrent.atomic.AtomicInteger;
 
 import static io.dapr.utils.TestUtils.assertThrowsDaprException;
```
Please use full imports

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -57,16 +56,10 @@
-import static org.junit.jupiter.api.Assertions.assertNotNull;
-import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.*;
 import static org.mockito.ArgumentMatchers.any;
```
Please use full imports

artur-ciocanu commented on sdk/src/test/java/io/dapr/client/DaprPreviewClientGrpcTest.java:
```diff
@@ -539,6 +532,285 @@ public void onError(RuntimeException exception) {
 	}
+
+
+	@Test
```
The formatting here seems off

artur-ciocanu replied:
@siri-varma could you please make sure you adjust the formatting of this file, something is really off here.

siri-varma replied:
done

artur-ciocanu commented on pom.xml:
```diff
@@ -17,7 +17,7 @@
     <grpc.version>1.69.0</grpc.version>
     <protobuf.version>3.25.5</protobuf.version>
     <protocCommand>protoc</protocCommand>
```
I think the latest is 1.15.3

artur-ciocanu commented on sdk-tests/src/test/java/io/dapr/it/testcontainers/DaprContainerConstants.java:
```diff
@@ -1,5 +1,5 @@
 package io.dapr.it.testcontainers;
 
 public interface DaprContainerConstants {
```
Same here

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/DaprClientImpl.java:
```diff
@@ -1290,6 +1296,144 @@ public Mono<UnsubscribeConfigurationResponse> unsubscribeConfiguration(Unsubscri
+  public Mono<Void> scheduleJob(ScheduleJobRequest scheduleJobRequest) {
+    try {
+      validateScheduleJobRequest(scheduleJobRequest);
+
```
I am wondering if we can move all the job request construction to a separate method so the `scheduleJob` reads like:
```
validateScheduleJobRequest(scheduleJobRequest)

grpcRequest = createGrpcScheduleJobRequest(scheduleJobRequest)

Mono<DaprProtos.ScheduleJobResponse> scheduleJobResponseMono =
          Mono.deferContextual(context -> this.createMono(
                  it -> intercept(context, asyncStub)
                      .scheduleJobAlpha1(DaprProtos.ScheduleJobRequest.newBuilder()
                          .setJob(scheduleJobRequestBuilder.build()).build(), it)
              )
          );
```

This way the main API is lean and clean.

siri-varma replied:
The function style in this PR follows the other method patterns in DaprClientBuilder. While I’m open to moving these into separate methods, I’m hesitant that it might create too many functions, making it tricky to understand the logic,

artur-ciocanu replied:
Agreed, let's leave it as is

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/DaprClientImpl.java:
```diff
@@ -1290,6 +1296,144 @@ public Mono<UnsubscribeConfigurationResponse> unsubscribeConfiguration(Unsubscri
+                          .setName(getJobRequest.getName()).build(), it)
+              )
+          );
+
```
Same here, we could move the GRPC response conversion to a separate method, so the main API method is clear

artur-ciocanu replied:
Let's leave it as is

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/DaprClientImpl.java:
```diff
@@ -1290,6 +1296,144 @@ public Mono<UnsubscribeConfigurationResponse> unsubscribeConfiguration(Unsubscri
+        scheduleJobRequestBuilder.setSchedule(scheduleJobRequest.getSchedule().getExpression());
+      }
+
+      if (scheduleJobRequest.getTtl() != null) {
```
Shouldn't we use some formatting here, I am not sure what is the default representation when invoking `toString` on offset date time

siri-varma replied:
https://docs.oracle.com/javase/8/docs/api/java/time/OffsetDateTime.html#toString--

toString method generates ISO8601 format which Dapr supports

artur-ciocanu replied:
As discussed let's use a well defined DateTimeFormatter instead of toString

artur-ciocanu replied:
As discussed let's use a well defined DateTimeFormatter instead of toString()

siri-varma replied:
Changed this to use Instant and added DateRimeFormatter

artur-ciocanu commented on sdk/src/main/java/io/dapr/client/domain/JobSchedule.java:
```diff
@@ -0,0 +1,119 @@
+ * JobsSchedule schedule = JobsSchedule.daily();
+ * System.out.println(schedule.getExpression()); // Outputs: "0 0 0 * * *"
+ * </pre>
+ */
```
I am wondering if this should be a simple cron expression holder and we should have a separate factory like `JobSchedules` that has the factory methods like `fromPeriod`, `early`, etc.

@cicoyle and @salaboy what are your thoughts.

salaboy replied:
@artur-ciocanu my only concern was to complicate things when we integrate with Spring, so I would keep it lighter on the pure java side and then reuse all the Spring helpers to create Job expressions.. does that make sense?

artur-ciocanu replied:
@salaboy my suggestion was to use `JobSchedule` as a simple class that wraps a `String` that is a CRON expression and we can have a helper factory like `JobScheduleFactory` that can have additional methods like `fromPeriod`, `fromString`, `early`, etc.

A CRON expression as a string is the lowest common denominator and it should be easy to integrate with Spring or any other framework.

Please let me know if I misunderstood your comment.

artur-ciocanu reviewed (commented):
@siri-varma I have left some comments. Could you please take a look and let me know your thoughts.

Thank you.

siri-varma commented on sdk/src/main/java/io/dapr/client/DaprClientImpl.java:
```diff
@@ -1290,6 +1296,144 @@ public Mono<UnsubscribeConfigurationResponse> unsubscribeConfiguration(Unsubscri
+      if (scheduleJobRequest.getRepeats() != null) {
+        scheduleJobRequestBuilder.setRepeats(scheduleJobRequest.getRepeats());
+      }
+
```
repeats is just an integer number which specifies number of times job should be scheduled.

artur-ciocanu replied:
I was referring to TTL and due time, both are OffsetDateTime.

artur-ciocanu commented:
@siri-varma could you please take a look at the tests, it seems that the scheduler container is not working as expected. Also I am wondering if we could avoid using `latest` tag in containers images, so we have more control when Dapr container group is upgraded.

artur-ciocanu commented:
@siri-varma I see the following errors for IT tests:
```
DaprComponentTest.containerConfigurationTest:42 » ContainerLaunch Container startup failed for image daprio/scheduler:latest
```
And
```
Container startup failed for image daprio/daprd:1.14.4
```
Could you please make sure you use the latest Dapr container `1.15.3` and ensure that the same image tag is used for all the other Dapr containers that are started behind the scenes, like scheduler or placement container.

siri-varma commented:
@artur-ciocanu I am traveling and will be back in couple of days. Will update the pr by Friday

siri-varma commented:
@artur-ciocanu Addressed all the comments, fixed integration tests

siri-varma commented:
The 1.14.4 is coming from actors. I will update it in a separate pr. Will create an issue

salaboy reviewed (approved):
This PR looks good to me.. @siri-varma did you wanted to make more changes? Is there something still not address from @artur-ciocanu comments?

siri-varma commented:
@salaboy Thanks for the review. Nope, I don't have anything else to address

artur-ciocanu commented:
@siri-varma could you please make sure that you do not accidentally replace fully qualified imports with `*`. Please make sure you configure your IDE do not collapse multiple imports into a single `*`.

Also I think it would be cleaner to wait for your PR #1299 and then fix the conflicts in this PR and then merge.

artur-ciocanu commented on testcontainers-dapr/src/main/java/io/dapr/testcontainers/DaprSchedulerContainer.java:
```diff
@@ -0,0 +1,82 @@
+
+/**
+ * Test container for Dapr scheduler service.
+ */
```
@siri-varma codecov is complaning about this file, could you please add a few unit tests to ensure the minimum functionality is covered, like port overrides, etc

artur-ciocanu reviewed (approved):
@siri-varma great job, I have left a few comments, could please take a look. Thank you.

artur-ciocanu commented:
@cicoyle I have reviewed the changes looks good, could you please take a look and approve. Thank you.

cicoyle commented on sdk-tests/src/test/java/io/dapr/it/testcontainers/DaprJobsIT.java:
```diff
@@ -0,0 +1,146 @@
+package io.dapr.it.testcontainers;
```
Please ensure the header is at the top of every file.

siri-varma replied:
Added headers

siri-varma commented:
@cicoyle Do you mind if I add the example, readme with markdown in a separate issue ? https://github.com/dapr/java-sdk/issues/1300. I have created one here and assigned it to myself.

Because this PR was already big thought doing it in a different PR would be better.

siri-varma commented:
> @siri-varma could you please make sure you adjust the formatting of this file, something is really off here.

@artur-ciocanu I did check the tab spaces and indentation and it looked fine. May be I am missing something ?

siri-varma commented:
I have added the Jobs sdk example with markdown in this commit https://github.com/dapr/java-sdk/pull/1255/commits/17429d645d8ea8f23d98724272fea12696e4ceda

Please take a look. Thank you

siri-varma commented:
@cicoyle I have added the Jobs sdk example with markdown in this commit https://github.com/dapr/java-sdk/pull/1255/commits/17429d645d8ea8f23d98724272fea12696e4ceda

Please take a look. Thank you

siri-varma commented:
@artur-ciocanu Please look at this commit to https://github.com/dapr/java-sdk/pull/1255/commits/68568c5f682381f999bda134dcd37a2e83afa5cb. I have reduced the spaces from 4 to 2

artur-ciocanu reviewed (approved):
Great work!

artur-ciocanu commented:
@cicoyle this PR looks good to be merged, please take a look and let's get it in. Thank you.

siri-varma commented:
@cicoyle let me know if I need to do anything else for merging the pr

cicoyle commented on examples/src/main/java/io/dapr/examples/jobs/README.md:
````diff
@@ -0,0 +1,123 @@
+}
+```
+
+Get into the examples' directory:
````
nit, can you remove this bc its already mentioned above? or remove it from here?

siri-varma replied:
Done removed the line

cicoyle commented on examples/src/main/java/io/dapr/examples/jobs/DemoJobsClient.java:
```diff
@@ -0,0 +1,49 @@
+
+    try (DaprPreviewClient client = new DaprClientBuilder().withPropertyOverrides(overrides).buildPreviewClient()) {
+
+      // Schedule a job.
```
mind adding a `system.out.println` saying 'Scheduling job <name>' or something along those lines? I ran this locally and it doesnt print anything which might be confusing for users - it'll be a bit clearer from terminal to see the log line.

siri-varma replied:
Done added system out printlins

cicoyle commented:
Thank you for getting the Jobs API over the line @siri-varma 🙌🏻
